src/Dependencies/Collections/Internal/SegmentedArrayHelper.cs

```diff
@@ -19,52 +19,70 @@ internal static class SegmentedArrayHelper
         [MethodImpl(MethodImplOptions.AggressiveInlining)]
         internal static int GetSegmentSize<T>()
         {
+            return Unsafe.SizeOf<T>() switch
+            {
+                // Hard code common values since not all versions of .NET support constant-folding the calculation.
+                // Values are validated against the reference implementation in CalculateSegmentSize in unit tests.
+                4 => 16384,
+                8 => 8192,
+                12 => 4096,
+                16 => 4096,
+                24 => 2048,
+                28 => 2048,
+                32 => 2048,
+                40 => 2048,
 #if NETCOREAPP3_0_OR_NEWER
-            return InlineCalculateSegmentSize(Unsafe.SizeOf<T>());
+                _ => InlineCalculateSegmentSize(Unsafe.SizeOf<T>()),
 #else
-            if (Unsafe.SizeOf<T>() == Unsafe.SizeOf<object>())
```

Comment:
>Unsafe.SizeOf<T>() == Unsafe.SizeOf<object>() [](http://example.com/codeflow?start=16&length=45)

Consider adding an assert verifying that this condition is always false.

Reply:
➡️ Renamed to `FallbackSegmentHelper<T>`